### PR TITLE
Combined Uniquify and LowerTypes pass 

### PR DESCRIPTION
**Pull request by ekiwi-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- _[n/a]_ Did you update the FIRRTL spec to include every new feature/behavior?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?

#### Type of Improvement
 - performance improvement
 - code refactoring

#### API Impact

- currently this just adds a `NewLowerTypes` pass
- we should see if we can just immediately replace `LowerTypes` with this new pass which should be doing strictly more

#### Backend Code Generation Impact

- this should not impact the generated Verilog
- an exception is the fact that we do better renaming and thus some annotation might not get dropped and thus affect the generated verilog

#### Desired Merge Strategy
- Rebase: You will rebase the PR onto master and it will be merged with a merge commit
- I will fix up the commit history before we merge

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by ekiwi-sifive** [comment]
I ran a cold compile benchmark and in general it seems to be slightly faster (but the SD is high as usual)


revision | design | max heap (MiB) | SD | runtime (s) | SD
-- | -- | -- | -- | -- | --
NewLowerTypes | small | 3055 | 98.14 | 48.12 | 3.05
master | small | 3065 | 126.98 | 51.44 | 1.49
NewLowerTypes | medium | 3590 | 85.04 | 112.55 | 2.89
master | medium | 3583 | 95.21 | 120.09 | 29.26

**Comment by ekiwi-sifive** [comment]
In terms of regression testing:

I compiled both the `small` and the `medium` benchmark with the current `master` branch and with the changes in this `PR` and the generated Verilog is exactly the same.

**Comment by ekiwi-sifive** [comment]
The only failing test now is the `LoweringCompilersSpec`

File: src/main/scala/firrtl/transforms/RemoveWires.scala [hunk 1]
```diff
@@ -23,16 +23,20 @@ import scala.util.{Try, Success, Failure}
...
+         Dependency(passes.ResolveKinds),
          Dependency(transforms.RemoveReset),
          Dependency[transforms.CheckCombLoops] )
 
   override def optionalPrerequisites = Seq(Dependency[checks.CheckResets])
 
   override def optionalPrerequisiteOf = Seq.empty
 
-  override def invalidates(a: Transform) = false
+  override def invalidates(a: Transform) = a match {
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/RemoveWires.scala:36]: Was this change necessary for this PR? Otherwise it should probably be separated into another PR.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+  private trait ParentRef { def ref(name: String, asVecField: Boolean = false): ReferenceTarget }
+  private case class ModuleParentRef(m: ModuleTarget) extends ParentRef {
+    override def ref(name: String, asVecField: Boolean): ReferenceTarget = m.ref(name)
+  }
+  private case class RefParentRef(r: ReferenceTarget) extends ParentRef {
+    override def ref(name: String, asVecField: Boolean): ReferenceTarget =
+      if(asVecField) { r.index(name.toInt) } else { r.field(name) }
+  }
+}
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:439]: newline

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+  */
+object NewLowerTypes extends Transform {
+  override def inputForm: CircuitForm = UnknownForm
+  override def outputForm: CircuitForm = UnknownForm
+
+  override def prerequisites: Seq[TransformDependency] = Seq(
+    Dependency(RemoveAccesses), // we require all SubAccess nodes to have been removed
+    Dependency(CheckTypes),     // we require all types to be correct
+    Dependency(InferTypes),     // we require instance types to be resolved (i.e., DefInstance.tpe != UnknownType)
+    Dependency(ExpandConnects)  // we require all PartialConnect nodes to have been expanded
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:27]: So does this work prior to expand whens?

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+    (newMod, renameMap, memInit)
+  }
+
+  def onPort(p: Port)(symbols: LoweringTable): Seq[Port] = symbols.lower(p)
+
+  def onStatement(s: Statement)(implicit symbols: LoweringTable, memInit: Seq[MemoryInitAnnotation]): Statement = s match {
+    // declarations
+    case d : DefWire =>
+      Block(symbols.lower(d.name, d.tpe, firrtl.WireKind).map{case (name, tpe, _) => d.copy(name=name, tpe=tpe) })
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:96]:
```suggestion
      Block(symbols.lower(d.name, d.tpe, firrtl.WireKind).map { case (name, tpe, _) => d.copy(name=name, tpe=tpe) })
```

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+    case d : DefMemory =>
+      // TODO: as an optimization, we could just skip ground type memories here.
+      //       This would require that we don't error in getReferences() but instead return the old reference.
+      val mems = symbols.lower(d)
+      if(mems.length > 1 && memInit.exists(_.target.ref == d.name)) {
+        val mod = memInit.find(_.target.ref == d.name).get.target.encapsulatingModule
+        val msg = s"[module $mod] Cannot initialize memory ${d.name} of non ground type ${d.dataType.serialize}"
+        throw new RuntimeException(msg)
+      }
+      Block(symbols.lower(d))
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:124]:
```suggestion
      Block(mems)
```

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+    case d : DefInstance => symbols.lower(d)
+    // connections
+    case Connect(info, loc, expr) =>
+      if(!expr.tpe.isInstanceOf[GroundType]) {
+        throw new RuntimeException(s"NewLowerTypes expects Connects to have been expanded! ${expr.tpe.serialize}")
+      }
+      val rhs = onExpression(expr)
+      // We can get multiple refs on the lhs because of ground-type memory ports like "clk" which can get duplicated.
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:132]: Huh, I'd forgotten about this. Nice catch!

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+      val lhs = symbols.getReferences(loc.asInstanceOf[RefLikeExpression])
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:133]: I wonder if it would be useful to change the `loc` type in `Connect` to be of `RefLikeExpression`. A future PR perhaps, this is fine as is.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+      Block(lhs.map(loc => IsInvalid(info, loc)))
+    // others
+    case other => other.mapExpr(onExpression).mapStmt(onStatement)
+  }
+
+  /** Replaces all Reference, SubIndex and SubField nodes with the updated references */
+  def onExpression(e: Expression)(implicit symbols: LoweringTable): Expression = e match {
+    case r: RefLikeExpression =>
+      // When reading (and not assigning to) an expression, we can always just pick the first one.
+      symbols.getReferences(r).head
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:152]: Can you update the comment to give an example of when getReferences returns a list of size greater than one? I imagine the memory's clock case is one example.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+    case other => other.mapExpr(onExpression)
+  }
+}
+
+// Holds the first level of the module-level namespace.
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:157]: What does "first level" mean? Please clarify this comment.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+  }
+
+  private def extractGroundTypeRefString(refs: Seq[ReferenceTarget]): String =
+  if(refs.isEmpty) { "" } else {
+    // Since we depend on ExpandConnects any reference we encounter will be of ground type
+    // and thus the one with the longest access path.
+    refs.reduceLeft((x,y) => if (x.component.length > y.component.length) x else y)
+    // convert references to strings relative to the module
+      .serialize.dropWhile(_ != '>').tail
+  }
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:359]:
```suggestion
    if(refs.isEmpty) { "" } else {
      // Since we depend on ExpandConnects any reference we encounter will be of ground type
      // and thus the one with the longest access path.
      refs.reduceLeft((x,y) => if (x.component.length > y.component.length) x else y)
      // convert references to strings relative to the module
        .serialize.dropWhile(_ != '>').tail
    }
```

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+
+  private def destruct(m: ModuleTarget, field: Field, rename: Option[RenameNode]): Seq[(Field, Seq[ReferenceTarget])] =
+    destruct(prefix = "", oldParent = ModuleParentRef(m), oldField = field, isVecField = false, rename = rename)
+
+
+  private def destruct(prefix: String, oldParent: ParentRef, oldField: Field,
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:365]: Some scaladoc would be helpful here to understand what's happening, especially in the larger context of where its called.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+  private def getFields(tpe: Type): Seq[Field] = tpe match {
+    case BundleType(fields) => fields
+    case v : VectorType => vecToBundle(v).fields
+  }
+
+  private def vecToBundle(v: VectorType): BundleType = {
+    BundleType(( 0 until v.size).map(i => Field(i.toString, Default, v.tpe)))
+  }
+
+  private trait ParentRef { def ref(name: String, asVecField: Boolean = false): ReferenceTarget }
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:431]: Please add some explanation of why these are necessary, and what they represent.

File: src/main/scala/firrtl/transforms/RemoveWires.scala (hunk 1 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/transforms/RemoveWires.scala:36]: I did that while debugging. I can move it to a separate PR.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala (hunk 3 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:27]: I haven't tested it but I haven't been able to find a fundamental reason why it shouldn't. As long as all references are lowered to ground types (ExpandConnects) and there are no SubAccess nodes, I think we should be fine.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala (hunk 5 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:124]: great catch!

File: src/main/scala/firrtl/passes/NewLowerTypes.scala (hunk 7 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:133]: I was trying to do that, but it messes with a lot of old code and requires you to insert quite some `asInstanceOf[]` casts. So yeah, maybe in the future.

File: src/main/scala/firrtl/transforms/RemoveWires.scala (hunk 1 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/transforms/RemoveWires.scala:36]: Now a separate PR: #1797

**Comment by ekiwi-sifive** [comment]
@seldridge , What exactly is the `Legacy Custom Transforms` behavior in `LoweringCompilerSpec` testing?
Is  it ok to add patches there? I am a little hesitant because so far no one else seems to have had to add patches.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+    destruct(prefix = "", oldParent = ModuleParentRef(m), oldField = field, isVecField = false, rename = rename)
+
+  /** Lowers a field into its ground type fields.
+    * @param prefix carries the prefix of the new ground type name
+    * @param isVecField is used to generate an appropriate old (field/index) reference
+    * @param rename The information from the `uniquify` function is consumed to appropriately rename generated fields.
+    * @return a sequence of ground type fields with new names and, for each field,
+    *         a sequence of old references that should to be renamed to point to the particular field
+    */
+  private def destruct(prefix: String, oldParent: ParentRef, oldField: Field,
```
Comment (azidar) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:376]: I'd add a comment that this function presumes the fields have already been uniquified.

**Comment by seldridge** [comment]
The `Legacy Custom Transforms` are testing where a custom transform, that uses `inputForm`/`outputForm` to specify dependencies, is placed in the transform sequence. Internally, `inputForm`/`outputForm` are being converted to dependencies and the resulting scheduling behavior needs to match how the old `getLoweringCompilers`/`mergeTransforms` methods worked. 

These tests aren't intended to be patched. (But they might need to be modified if in-tree transforms start specifying very sparse dependencies.)

By example: if you have `inputForm=LowForm`/`outputForm=MidForm`, you'd expect to see the transform run after low form optimizations, and then re-run from mid form to low form optimizations. [The test manually constructs this order and then compares it against the order if you just requested low form optimizations and the custom transform as targets.](https://github.com/freechipsproject/firrtl/blob/master/src/test/scala/firrtlTests/LoweringCompilersSpec.scala#L316)

This is mostly straightforward except for the edge cases identified in: https://github.com/freechipsproject/firrtl/issues/1165. The `outputForm=HighForm` edge cases are tested in `LoweringCompilerSpec.scala`. The edge cases of `inputForm=LowForm` are tested in `CustomTransformSpec.scala`.

File: src/test/scala/firrtl/passes/LowerTypesSpec.scala [hunk 14]
```diff
@@ -0,0 +1,538 @@
...
+    val renames = RenameMap()
+    val mutableSet = scala.collection.mutable.HashSet[String]() ++ namespace
+    val res = DestructTypes.destruct(m, ref, mutableSet, renames)
+    DestructResult(resultToFieldSeq(res), renames)
+  }
+  def resultToFieldSeq(res: Seq[(firrtl.ir.Field, String)]): Seq[String] =
+    res.map(_._1).map(r => s"${r.flip.serialize}${r.name} : ${r.tpe.serialize}")
+  def get(r: RenameMap, m: IsMember): Set[IsMember] = r.get(m).get.toSet
+  protected val m = CircuitTarget("m").module("m")
+}
```
Comment (azidar) [inline src/test/scala/firrtl/passes/LowerTypesSpec.scala:538]: newline

File: src/test/scala/firrtlTests/LoweringCompilersSpec.scala [hunk 15]
```diff
@@ -120,6 +120,8 @@ class LoweringCompilersSpec extends FlatSpec with Matchers {
 
     val patched = scala.collection.immutable.TreeMap(m.toArray:_*).values.flatten
 
+    println(patched.map(_.getName).mkString("\n"))
```
Comment (azidar) [inline src/test/scala/firrtlTests/LoweringCompilersSpec.scala:123]: I presume this println will get removed after the tests pass?

**Comment by ekiwi-sifive** [comment]
Thanks for the detailed explanation Schuyler!

> resulting scheduling behavior needs to match how the old `getLoweringCompilers`/`mergeTransforms` methods worked.

I combined `Uniquify` and `LowerTypes` into a single `NewLowerTypes` pass. How should that be reflected in the tests?

File: src/test/scala/firrtlTests/LoweringCompilersSpec.scala (hunk 15 above)
Comment (ekiwi-sifive) [reply src/test/scala/firrtlTests/LoweringCompilersSpec.scala:123]: Good find! This was never supposed to be committed!

File: src/main/scala/firrtl/passes/NewLowerTypes.scala (hunk 13 above)
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:376]: The fields haven't really been uniquified quite yet. The uniquifying information has been calculated, but we actually apply it in this function (which is what the `rename` parameter is for).
Is there a way I could clarify that?

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+import scala.collection.mutable
+
+/** Flattens Bundles and Vecs.
+  * - Some implicit bundle types remain, but with a limited depth:
+  *   - the type of a memory is still a bundle with depth 2 (mem -> port -> field), see [[MemPortUtils.memType]]
+  *   - the type of a module instance is still a bundle with depth 1 (instance -> port)
+  */
+object NewLowerTypes extends Transform {
+  override def inputForm: CircuitForm = UnknownForm
+  override def outputForm: CircuitForm = UnknownForm
```
Comment (seldridge) [inline src/main/scala/firrtl/passes/NewLowerTypes.scala:21]: Mixin `with DependencyAPIMigration` and remove deprecated `inputForm`/`outputForm`.

Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:21]: Will do. I didn't do this initially since the word *Migration* suggested to me that this is only for migrating legacy passes.

**Comment by seldridge** [comment]
The custom transform `inputForm=MidForm`/`outputForm=HighForm` test is doing something weird, but this looks like it's a problem with the `DependencyManager`.

Specifically, after the custom transform runs, the state is high form. `NewLowerTypes` is scheduled to run next so there's a state mismatch. The dependency manager solves a sub-problem to meet the prerequisites of `NewLowerTypes`. However, that is a minimal set of four transforms. The `DependencyManager` then thinks that it's only added those four transforms to the state, but it's actually added a ton of other stuff, like `ToWorkingIR`.

Then, after `NewLowerTypes` runs, the dependency manager re-runs `ToWorkingIR` and a ton of other stuff.

In summary, the `targets` for a dependency manager is the minimal set of things that it does, but it may actually do more work.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 17]
```diff
@@ -0,0 +1,466 @@
...
+import firrtl.stage.TransformManager.TransformDependency
+
+import scala.collection.mutable
+
+/** Flattens Bundles and Vecs.
+  * - Some implicit bundle types remain, but with a limited depth:
+  *   - the type of a memory is still a bundle with depth 2 (mem -> port -> field), see [[MemPortUtils.memType]]
+  *   - the type of a module instance is still a bundle with depth 1 (instance -> port)
+  */
+object NewLowerTypes extends Transform with DependencyAPIMigration {
```
[the remark this answers is not in the document]
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:19]: The plan was to eventually just name it `LowerTypes` after we make sure that it provides the same public methods as `LowerTypes`.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+  def loweredName(s: Seq[String]): String = s.mkString(delim)
+
+  override def execute(state: CircuitState): CircuitState = {
+    // When memories are lowered to ground type, we have to fix the init annotation or error on it.
+    val (memInitAnnos, otherAnnos) = state.annotations.partition {
+      case _: MemoryRandomInitAnnotation => false
+      case _: MemoryInitAnnotation => true
+      case _ => false
+    }
+    val memInitByModule = memInitAnnos.map(_.asInstanceOf[MemoryInitAnnotation]).groupBy(_.target.encapsulatingModule)
```
[the remark this answers is not in the document]
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:53]: Currently we need to know these annotations in order to throw an error if they are applied to a non-ground type memory (this is a feature that was already in the old `LowerTypes` pass).
I was planning to add custom logic to make these annotations work with non-ground type memories, but that does require me to have information about all the memories that the original memory gets split up into. Not sure how I would get the information in the `duplicate` method of the annotation.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 19]
```diff
@@ -0,0 +1,466 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl.analyses.InstanceKeyGraph
+import firrtl.annotations.{CircuitTarget, MemoryInitAnnotation, MemoryRandomInitAnnotation, ModuleTarget, ReferenceTarget}
+import firrtl.{CircuitForm, CircuitState, DependencyAPIMigration, InstanceKind, Kind, MemKind, PortKind, RenameMap, SymbolTable, Transform, UnknownForm, Utils}
```
[the remark this answers is not in the document]
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:7]: I wanted to be verbose about it and its not much of a problem with an IDE. If you want, I can change it to `import firrtl._`.

File: src/main/scala/firrtl/passes/NewLowerTypes.scala [hunk 20]
```diff
@@ -0,0 +1,466 @@
...
+  /** Expands a chain of referential [[firrtl.ir.Expression]]s into the equivalent lowered name
+    * @param e [[firrtl.ir.Expression]] made up of _only_ [[firrtl.WRef]], [[firrtl.WSubField]], and [[firrtl.WSubIndex]]
+    * @return Lowered name of e
+    * @note Please make sure that there will be no name collisions when you use this outside of the context of LowerTypes!
+    */
+  def loweredName(e: Expression): String = e match {
+    case e: Reference => e.name
+    case e: SubField => s"${loweredName(e.expr)}$delim${e.name}"
+    case e: SubIndex => s"${loweredName(e.expr)}$delim${e.value}"
+  }
```
[the remark this answers is not in the document]
Comment (ekiwi-sifive) [reply src/main/scala/firrtl/passes/NewLowerTypes.scala:43]: I copied this function from `LowerTypes` because it is used by other passes, so I cannot change the argument type, since I want to try and replace `LowerTypes` without a deprecation cycle. However I think it is a good idea to add a better error message than a match error.

**Comment by ekiwi-sifive** [comment]
> @seldridge this is a good example where external transforms that depend on `Uniquify` or `LowerTypes` may have awkward interactions (although I guess it should at least "work").

The plan is to delete the old `LowerTypes` transform and rename the `NewLowerTypes` to just `RenameTypes`. I just haven't done that yet because I need to go through all the public fields of `LowerTypes` and make sure we provide compatible replacements. `Uniquify` will stay around and be deprecated. So now if you have a custom transform that only depends on `Uniquify`, the deprecated version of `Uniquify` will be run. If your transform depends on both, `LowerTypes` and `Uniquify`, the run of `Uniquify` will be redundant, but in general nothing should break if the new version of `LowerTypes` tries to uniquify again.

Let me know if that transition plan sounds reasonable to you.

**Comment by seldridge** [comment]
@jackkoenig wrote:
> @seldridge this is a good example where external transforms that depend on Uniquify or LowerTypes may have awkward interactions (although I guess it should at least "work"). How hard do you think it would be to support an implements or provides API?

I don't think this is too bad and it's probably the way that we need to go in the mid-term. This is also a mechanism to wholly replace in-tree transforms with custom implementations.

I think this would look like a special transform which, after application, causes a "property" to be applied to the circuit. Dependencies can then be specified either as explicit transforms or as `Provider[A <: Property]`. If it's only requesting a property, then the dependency manager looks around to try and find all transforms that can provide the property. There's then some API for providing an ordering of providers. The dependency manager then either succeeds and will always choose one provider for that property. If it fails to find a provider then it fails. Somewhat annoyingly, I expect this requires using META-INF which isn't amazing.

tl;dr: I think we can do this with no modification to the actual dependency API, just with a new mix-in.

This would be a separate endeavor from this PR.

**Comment by ekiwi-sifive** [comment]
So right now there are four public methods of the old `LowerTypes` pass that I do not intend to provide with the `NewLowerTypes` pass:

```.scala
  def renameExps(renames: RenameMap, n: String, t: Type, root: String): Seq[String]
  def renameExps(renames: RenameMap, n: String, t: Type): Seq[String]
  def renameExps(renames: RenameMap, e: Expression, root: String): Seq[String]
  def lowerTypes(renames: RenameMap, initializedMems: Set[(String, String)])(m: DefModule): DefModule
```

They are never used outside the `LowerTypes` pass in the `firrtl` code base and were probably made public by accident.

Would it be acceptable to drop these methods without a deprecation cycle? This would allow us to just overwrite the old `LowerTypes` pass with the new one.

**Comment by ekiwi-sifive** [comment]
The only test failing now is `LoweringCompilerSpec`, in particular `Legacy Custom Transforms` for `Mid -> High` and `Mid -> Chirrtl`. I think I will need @seldridge 's help to fix that.

**Comment by azidar** [comment]
LGTM.
